rfcs/text/0352-graph-type-system.md

```diff
@@ -0,0 +1,2659 @@
+}
+```
+
+More specifically, it's required to allow `oneOf` with more variants than _just Data Types_ because:
+
+```json
+{
+  "someProperty": [{ "bar": 2.0 }, "foo"]
+}
+```
+
+Requires something like
+
+```json
+{
+  "name": "someProperty",
+  "oneOf": [
+    {
+      "type": "object",
+      "properties": {
+        "bar": { "$ref": "property-type/bar" }
+      }
+    },
+    { "$ref": "data-type/text" }
+  ]
+}
+```
+
+### Including an `Empty List` Data Type
+
+Similarly, the `Empty List` Data Type is required to be able to programmatically map unknown JSON data into the system. Given an incoming entity like:
+
+```json
+{
+    "foo": [
+        [],
+        [2, 6, 10, ...],
+        [{...}, {...}, ...]
+    ]
+}
+```
+
+There isn't a good assumption to make about the type of the first array. As such, including `Empty List` makes it possible to treat this as a value without an inner type, and a user will be forced to retype it to be able to modify it.
+
+### Including an `Object` Data Type
+
+The inclusion of the `Object` Data Type is less strongly supported. At the moment it seems like it's easy to include in implementations (as datastores will already need to be able to represent JSON-like structures), so instead the discussion at the moment is around looking for a reason to exclude it.
+
+There's a risk that it could be utilized instead of Property Type objects, which means the inner data won't be accessible through tooling designed around the Type System. However at the moment it seems a convenient inclusion to allow users to prototype and quickly throw in semi-structured data as black boxes, later on transitioning to Data Types. As the type is very constrained in the current proposal, with it not being possible to add further JSON schema keywords, etc. the risk that users will misuse it seems low.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+As mentioned in a few sections, this design basically defines a way for communities to build a compound ontology. As such it's worth mentioning some of the alternative knowledge representation models to provide points of discussion, and to invite members of their communities to give thoughts. There are a few specific technologies that we've taken particular interest in while designing the RFC.
+
+> Note: Some planning has been made around **varying**-levels of compatibility with these technologies, although in-depth designs for their interfacing have not been written up yet.
+
+## Data Description Formats
+
+- [RDF](https://www.w3.org/RDF/)
+- [RDFS](https://www.w3.org/TR/rdf-schema/)
+- [OWL](https://www.w3.org/OWL/)
+
+## Ontologies
+
+- [Schema.org](https://schema.org/)
+- [DBPedia](https://dbpedia.org/)
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+1.  Are there further impacts on block schemas?
+1.  Do we permit spaces in type names?
+1.  Is there a way to specify in JSON schema that the key of a property is equal to the the thing it's a `$ref` to?
```

Comment:
It would help to have an explanation somewhere that we want this / why is it desirable

Reply:
I've added some technical reason why (https://github.com/blockprotocol/blockprotocol/pull/352/commits/7842a041d3cdd435776db3194e097ee1c4073d86), we might also have some spec-reason why this is desirable

Reply:
Thank you – this addresses why we want to be able to _validate_ this, although we still don't explain anywhere why we want them to be the same at all.

We can track the need for that in this [thread](https://github.com/blockprotocol/blockprotocol/pull/352#discussion_r910774375), so I'll resolve this.